### PR TITLE
Opt-in nukies

### PR DESCRIPTION
**Pull request by Veritius**
## About the PR 
Adds the ability for players to opt into being nukies.

**Changelog**

:cl:
- fix: You can now opt in to being a nukie.

### CODE REVIEW

File: Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs [hunk 1]
```diff
@@ -91,17 +103,53 @@ private void OnPlayersSpawning(RulePlayerSpawningEvent ev)
...
+            else
+            {
+                nukeOp = _random.PickAndTake(prefList);
+                list.Remove(nukeOp);
+                Logger.InfoS("preset", "Selected a preferred nukeop.");
+            }
+            operatives.Add(nukeOp);
+        }
+
+        var map = _config.Shuttles != null ? _random.Pick(_config.Shuttles) : "Maps/infiltrator.yml";
```
Comment (PaulRitter) [inline Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs:152]: just give the datafield a default value with this map instead of doing that logic here

**Review by PaulRitter — CHANGES_REQUESTED** [review]
not a big fan of the gameruleconfig implementation but i cant come up with anything better atm so i'll let it slide

**Comment by Veritius** [comment]
sleep deprived coding moment
i'll fix it tomorrow

File: Content.Server/GameTicking/Rules/Configurations/GenericGameRuleConfiguration.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+/// <summary>
+/// A generic configuration, for game rules that don't have special config data.
+/// </summary>
+[UsedImplicitly]
+public sealed class GenericGameRuleConfiguration : GameRuleConfiguration
+{
+    [DataField("id", required: true)]
+    private string _id = default!;
+    public override string Id => _id;
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/Configurations/GenericGameRuleConfiguration.cs:13]: wut

File: Content.Server/GameTicking/Rules/Configurations/NukeopsGameRulePrototype.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    };
+
+    /// <summary>
+    /// Loadouts for regular operatives
+    /// </summary>
+    [DataField("operativeLoadouts")]
+    public List<string> OperativeLoadouts = new()
+    {
+        "SyndicateOperativeGearFull"
+    };
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/Configurations/NukeopsGameRulePrototype.cs:52]: These all need typeserializers.

**Comment by metalgearsloth** [comment]
> not a big fan of the gameruleconfig implementation but i cant come up with anything better atm so i'll let it slide

I don't think it offers anything over just using an IPrototype directly tbh.

File: Content.Server/GameTicking/Rules/GameRulePrototype.cs [hunk 4]
```diff
@@ -1,10 +1,14 @@
...
 using Robust.Shared.Prototypes;
 
 namespace Content.Server.GameTicking.Rules;
 
 [Prototype("gameRule")]
 public sealed class GameRulePrototype : IPrototype
 {
-    [IdDataFieldAttribute]
+    [IdDataField]
     public string ID { get; } = default!;
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/GameRulePrototype.cs:10]: ?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
How gamerules are used is kinda mixed. Nukies hardcodes the prototype but something else uses it in the method but then something else doesn't even use it. This should be cleaned up.

This also leaves the cvars hanging around when they should be merged as well.

The nukies defaults also don't match the prototype for reasons.

File: Content.Server/GameTicking/Rules/GameRulePrototype.cs (hunk 4 above)
Comment (Veritius) [reply Content.Server/GameTicking/Rules/GameRulePrototype.cs:10]: i have no goddamn idea this is all the work of #8539

**Comment by Veritius** [comment]
> > not a big fan of the gameruleconfig implementation but i cant come up with anything better atm so i'll let it slide
> 
> I don't think it offers anything over just using an IPrototype directly tbh.

this discussion should probably be moved to #8539 since that's the pr that adds the stuff you're discussing

File: Content.Server/GameTicking/Rules/Configurations/GenericGameRuleConfiguration.cs (hunk 2 above)
Comment (moonheart08) [reply Content.Server/GameTicking/Rules/Configurations/GenericGameRuleConfiguration.cs:13]: tl;dr I didn't want to make a new class for every single game rule that doesn't already have a config, and you can't override fields.

File: Content.Server/GameTicking/Rules/GameRulePrototype.cs (hunk 4 above)
Comment (metalgearsloth) [reply Content.Server/GameTicking/Rules/GameRulePrototype.cs:10]: Aight will move over there.

File: Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs [hunk 5]
```diff
@@ -91,16 +96,90 @@ private void OnPlayersSpawning(RulePlayerSpawningEvent ev)
...
+                    }
+                    else
+                    {
+                        nukeOp = _random.PickAndTake(prefList);
+                        Logger.InfoS("preset", "Insufficient preferred nukeop commanders, picking at random from regular op list.");
+                    }
+                }
+                else
+                {
+                    nukeOp = _random.PickAndTake(cmdrPrefList);
```
Comment (metalgearsloth) [inline Content.Server/GameTicking/Rules/NukeopsRuleSystem.cs:156]: Shouldn't this remove them from the everyone + prefList as well in case of duplicates?

**Pull request merged**